### PR TITLE
Psalm types

### PR DESCRIPTION
**Pull request by DustinS2S**
@flumonion Please review, the only one I wasn't sure of was the `meta`

### CODE REVIEW

**Comment by DustinS2S** [comment]
Removing `[dateStringFrom for created and modified](https://github.com/stock2shop/share/pull/32/commits/8bcd9ab38a0e46dc5e6ee45e071bb00ca2e4df99)` commit

File: src/DTO/Address.php [hunk 1]
```diff
@@ -6,6 +6,24 @@
 
 use JsonSerializable;
 
+/**
+ * @psalm-type ptAddress = array{
```
Comment (flumonion) [inline src/DTO/Address.php:10]: Why `ptAddress` and not `Address`?
i.e. match class name? those are unique.

File: src/DTO/Channel.php [hunk 2]
```diff
@@ -6,6 +6,22 @@
...
 use JsonSerializable;
 
+/**
+ * @psalm-type ptChannel = array{
+ *     active: bool,
+ *     client_id: int,
+ *     created: string,
+ *     description: string,
+ *     id: int,
+ *     meta: array,
```
Comment (flumonion) [inline src/DTO/Channel.php:16]: You need to define the meta.
In the Meta Class, create the Psalm data type for meta.
Then here, you need to define an array of that type.

**Review by flumonion — CHANGES_REQUESTED** [review]

File: src/DTO/Channel.php (hunk 2 above)
Comment (flumonion) [reply src/DTO/Channel.php:16]: I forgot to mention, you can `import` the data type from the other class.
https://psalm.dev/docs/annotating_code/supported_annotations/#psalm-import-type

File: src/DTO/Address.php (hunk 1 above)
Comment (DustinS2S) [reply src/DTO/Address.php:10]: @flumonion I renamed them to be slightly different to the class as psalm appears to fail when it is named the same, see below when running psalm:
```
ERROR: ImplementedReturnTypeMismatch - src/DTO/ChannelImage.php:56:16 - The inherited return type 'array<array-key, Stock2Shop\Share\DTO\Image>' 
for Stock2Shop\Share\DTO\Image::createArray is different to the implemented return type for Stock2Shop\Share\DTO\ChannelImage::createarray 'array<array-key, array{active: bool, channel_id: int, channel_image_code: string, delete: bool, id: int, success: bool}>' (see https://psalm.dev/123)
     * @return ChannelImage[]
```
The issue is related to the `@return` for the `createArray` function, psalm seems to get confused when the return type is the same as the psalm-type.

To resolve this I can either:
1. Use unique psalm-type names
2. Change the `@return` to be `self[]` on all the `createArray` functions

I am not sure which to suggest as I am not sure if using the same name as the class name may cause issues when the class is called in any other PHP prams docs

Comment (flumonion) [reply src/DTO/Address.php:10]: `TypeAddress`, `TypeChannel` etc is fine...
`pt` is just a bit ambiguous :)

**Comment by DustinS2S** [comment]
@flumonion Please review.
Please do not merge, if you are happy with the types, I will do all the types on this branch and request the merge when ready.

File: src/DTO/Address.php [hunk 3]
```diff
@@ -6,6 +6,24 @@
 
 use JsonSerializable;
 
+/**
+ * @psalm-type TypeAddress = array{
+ *     address1: string,
```
Comment (flumonion) [inline src/DTO/Address.php:11]: I think you are missing the `?` for optional array properties.
https://psalm.dev/docs/annotating_code/type_syntax/array_types/#array-shapes

A DTO can be constructed without any of these properties, there may be some edge cases.

**Comment by DustinS2S** [comment]
@flumonion 
Please could you review and pay special attention to the following files:
`ChannelOrder`
`ChannelOrderCustomer.php`

I am not sure how to declare the type when the class extends a parent class and has variables like the above two files.

**Comment by flumonion** [comment]
I don't think you can extend Psalm data types the same way you do with classes?
If this is the case, then you need to write out all the possible properties for each data type that extends some other data type.
I could be wrong...

**Comment by DustinS2S** [comment]
@flumonion Ok, I will do that and do some reading on the topic and revert here.

**Comment by DustinS2S** [comment]
@flumonion 
It appears there is no way to prevent Psalm from complaining when a child class constructor calls the parent class constructor.
The error:
```
ERROR: InvalidArgument - src/DTO/ChannelImage.php:36:29 - Argument 1 of Stock2Shop\Share\DTO\Image::__construct expects array{src?: string}, but array{active?: bool, channel_id?: int, channel_image_code?: string, delete?: bool, id?: int, src?: string, success?: bool} with additional array shape fields (active, channel_id, channel_image_code, delete, id, success) was provided (see https://psalm.dev/004)
        parent::__construct($data);
``` 

The only two possible solutions I could find:
1. Add all properties from child classes to parent class 
2. Use inline tag `@psalm-suppress InvalidArgument` where ever a parent class constructor is called

I have used the second option for now, please let me know if there is something else I can do or when you would prefer?

**Comment by DustinS2S** [comment]
@flumonion Updated, please review

**Pull request merged**